### ISSUE TITLE
Hotkey to quickly switch to between running apps as shown in the panel

### ISSUE THREAD

**Issue by stevep99** [home-sweet-gnome/dash-to-panel#1736]
**Describe the idea or feature you'd like**
New configurable hotkeys would be added that would perform task switching ("next-app" and "previous-app") in a fast and efficient way. This would be similar to the method used by [material shell](https://material-shell.com/).

Consider the current typical way to do this: 
1. Press Alt-TAB (or sometimes Super-TAB). A popup appears. [Note that the order of the icons in this popup is unrelated to the to the currently running apps as shown in dash-to-panel.]
2. While holding Alt, keep tapping TAB until until the desired app is highlighted.
3. Release TAB.

Now compare this to other task switching mechanisms, such as that in material shell, where there are simple and intuitive arrow-based shortcuts (e.g. Super-A, Super-D). These shortcuts immediately switch to the next/previous app in a predictable way, consistent with the on-screen visualisation of currently running apps. This is easily a superior method, and it would be brilliant if dash-to-panel supported it. NB I am only talking about task switching mechanic here, NOT workspace switching. There are already good GNOME shortcuts for that, so it's out of scope of this ticket. NB also, no UI changes to the panel are expected for this.

So, what I am asking for, is the following:

_Given_ you have three apps open
And these three apps are underlined in the panel (call them A, B, C in that order)
And A is the current focussed (highlighted) app.
_Then_ pressing the proposed "next-app" shortcut brings app B to the foreground (and B will be highlighted in the panel)
And pressing the proposed "next-app" shortcut again, brings app C to the foreground (and C will be highlighted in the panel)

(And of course the "previous-app" shortcut would switch in the reverse direction)

**Is your feature request related to a problem? Please describe.**
I am unsatisfied with the existing standard GNOME task switching system, but think by combining task switching with dash-to-panel's existing on-screen visualisation of running tasks, would be the perfect combo.

**Comment by tannerlegvold** [comment]
I would like this feature too. I use [Material Shell](https://material-shell.com/) but literally only for this feature (and the vertical workspaces). The workspaces already keep track of the focused window which is the hard part. 

I can almost implement this myself. With [Activate Window By Title](https://github.com/lucaswerkmeister/activate-window-by-title) installed you can execute
```
gdbus call --session \
    --dest org.gnome.Shell \
    --object-path /de/lucaswerkmeister/ActivateWindowByTitle \
    --method de.lucaswerkmeister.ActivateWindowByTitle.activateBySubstring \
    'Firefox'
```
in a terminal to have a window with `Firefox` in the name get focused (only works for windows in the current workspace, which is fine for us, but could probably be fixed). At least for me, the Dash to Panel dashboard gets updated correctly.

With [window-calls](https://github.com/ickyicky/window-calls) and [`jq`](https://github.com/stedolan/jq) installed
```
gdbus call --session --dest org.gnome.Shell --object-path \
/org/gnome/Shell/Extensions/Windows --method \
org.gnome.Shell.Extensions.Windows.List | tail -c +3 | head -c -4 | jq '.'
```
we can get a nice JSON object of all sorts of window related data. Problem: this gives all windows across all workspaces in no particular order. What we need is the windows in this workspace in the order Dash to Panel displays them. I'm sure Dash to Panel's code has a nice variable for this somewhere, but I'm not keen to go hunting. Hopefully one of the developers can help me out here.

I'm not a Gnome programmer, but if I had that variable I think I could work out the rest of it.

**Comment by tannerlegvold** [comment]
bump

**Comment by tannerlegvold** [comment]
I've been working on this. Heres where I'm at: by studying the source code for an amount of time I'm not willing to admit, I pieced together this
```
windows = global.dashToPanel.panels[0].taskbar.getAppInfos().reduce((ws, appInfo) => ws.concat(appInfo.windows), [])
```
which, as long as the Dash to Panel extension is enabled, seems to return a list of the windows in the order of the taskbar. You can test this by copying the above code hitting Alt + F2 and typing `lg` then Enter to bring up Looking Glass, paste the code in and hit Enter. For example (in Looking Glass) now run `windows.map((win) => win.get_title())` you should get a list of the titles of the windows.

This can be called in Bash with eg
```
gdbus call \
  --session \
  --dest org.gnome.Shell \
  --object-path /org/gnome/Shell \
  --method org.gnome.Shell.Eval "
    windows = global.dashToPanel.panels[0].taskbar.getAppInfos().reduce((ws, appInfo) => ws.concat(appInfo.windows), []); 
    windows.map((win) => win.get_title())"
```
if you've turned on Unsafe mode on your system which can be done with [this Unsafe Mode Menu](https://github.com/linushdot/unsafe-mode-menu) extension.

With this one could implement @stevep99's requested feature in Bash using Gnome's builtin (through Settings that is) keybinding mechanism. However, we could also implement the keybinding through code in an extension, either a fork of Dash to Panel or (probably better) another extension meant to go on top of it. Its easy to write a `focus taskbar sibling` function, [see here](https://github.com/home-sweet-gnome/dash-to-panel/blob/master/utils.js#L385), in GJS. 

Unfortunately keybindings in extensions take some setup (edits to schema files and so on) that I don't wanna bother with right now.

Of course now I'm wondering how to swap/reorder two windows in the taskbar or move a window to a different workspace in the code.

**Comment by charlesg99** [comment]
There already is such a mechanism with the mouse scroll on the panel.

**Comment by tannerlegvold** [comment]
> I understand that you'd like a keyboard shortcut
Don't worry, by now I'm not requesting any of this stuff be added. At this point I'm just using this thread as some personal notes, since this is functionality I, but apparently few others, want. Sorry about the pings.

> on a multi-monitor setup you can't select which panel applications to cycle
I pretty much only use a laptop so I've got no experience with multi-monitor window organization, but if the problem is ambiguity as to which monitor the user wants the keybinding to apply to, I'd think a good answer is just which ever has the currently focused window (is this what primary monitor means?). If I type a `j` on my keyboard it's not like theres ambiguity as to where it will go, it will go where ever the blinking cursor is, should putting <Super> in front change that?

While I've got you here, is there an easy way to swap/reorder two windows next to each other in the taskbar?

**Comment by charlesg99** [comment]
You're right that we could simply use the monitor where the currently focused window is and fallback to the primary panel if there are none (or no panel on this monitor). That would make sense without adding too many settings! I already thought about adding a generic keyboard shortcut "editor" that would allow to launch custom actions (that we could add overtime as new features). That would be a good candidate to start with now that the adwaita dialog is in place.

For the reordering, if the applications are ungrouped and your taskbar is unlocked, you can drag the icons to reorder the windows. If they're ungrouped only, you can reorder the windows within their application group. To mix windows from different application, you need to use the "launcher mode" or hide the favorites from your taskbar. Thanks!

**Comment by tannerlegvold** [comment]
Ok I went ahead and implemented it [as another extension](https://github.com/tannerlegvold/hotkeysForActivatingSiblingsInDashToPanel/blob/main/README.md). I only did activation of neighboring windows. No swapping yet. I like it. Perhaps it will be of use to others who find this thread :)

Also just today I found [issue 1053](https://github.com/home-sweet-gnome/dash-to-panel/issues/1053) which requested the exact same thing 3 years ago. Someone there also implemented it, that time [as a a fork of Dash to Panel](https://github.com/philippotto/dash-to-panel). I tried it but it didn't work. Wasn't sure if it installed wrong or just doesn't work after 3 years of bit rot.

**Comment by stevep99** [comment]
@tannerlegvold  Tried it out and it seems to be working well. I just needed to change the metadata file so it would work with Gnome 43. Will keep it installed and see how it goes. Nice work!

**Comment by charlesg99** [comment]
Great!

**Comment by RanzQ** [comment]
I improved this extension by @tannerlegvold, in case someone would find it useful:
https://github.com/RanzQ/hotkeysForActivatingSiblingsInDashToPanel/tree/jump-to-group

Fixed it for newer Gnome and instead of the planned "swapping", I added "jump to group", which works kind of like alt-tab but I get to navigate the apps taskbar order. I've always hated the "most recent" listing in Gnome which prevents quick navigation.

Also I think it could be useful to optionally make this "cycle windows" to be per app but didn't implement it yet.
